From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tariq Bashir <tbashir@brightsign.biz>
Date: Tue, 2 Dec 2025 16:01:17 +0000
Subject: Enable memory pressure monitoring and enable for renderer process
 OS-20368

Chromium has disabled memory pressure monitoring for Linux by default
(https://chromium-review.googlesource.com/c/chromium/src/+/2792266). This patch
re-enables it and also cherry picks the QtWebEngine patch that enables memory pressure
monitoring for the Linux renderer processes from:
https://cam-gerrit.brightsign.info/c/qtwebengine-chromium/+/15831

diff --git a/components/memory_pressure/system_memory_pressure_evaluator.cc b/components/memory_pressure/system_memory_pressure_evaluator.cc
index 3299c7fab7899d8d18e590addb66a4cd1900adfe..3a74620122bb2bc4491c3b85b143b5a21ec0172d 100644
--- a/components/memory_pressure/system_memory_pressure_evaluator.cc
+++ b/components/memory_pressure/system_memory_pressure_evaluator.cc
@@ -17,6 +17,8 @@
 #include "components/memory_pressure/system_memory_pressure_evaluator_mac.h"
 #elif BUILDFLAG(IS_WIN)
 #include "components/memory_pressure/system_memory_pressure_evaluator_win.h"
+#elif BUILDFLAG(IS_LINUX)
+#include "components/memory_pressure/system_memory_pressure_evaluator_linux.h"
 #endif
 
 namespace memory_pressure {
@@ -51,6 +53,9 @@ SystemMemoryPressureEvaluator::CreateDefaultSystemEvaluator(
     evaluator->CreateOSSignalPressureEvaluator(monitor->CreateVoter());
   }
   return evaluator;
+#elif BUILDFLAG(IS_LINUX)
+  return std::make_unique<memory_pressure::os_linux::SystemMemoryPressureEvaluator>(
+      monitor->CreateVoter());
 #else
   // Chrome OS and Chromecast evaluators are created in separate components.
   return nullptr;
diff --git a/content/renderer/render_thread_impl.cc b/content/renderer/render_thread_impl.cc
index 7b4fb3da23f4e7805f74932cd959229554c3e217..c65b533af1f980bcb2c17d9df5196255b30165ed 100644
--- a/content/renderer/render_thread_impl.cc
+++ b/content/renderer/render_thread_impl.cc
@@ -63,6 +63,7 @@
 #include "cc/trees/raster_context_provider_wrapper.h"
 #include "cc/trees/ukm_manager.h"
 #include "components/discardable_memory/client/client_discardable_shared_memory_manager.h"
+#include "components/memory_pressure/multi_source_memory_pressure_monitor.h"
 #include "components/metrics/public/mojom/single_sample_metrics.mojom.h"
 #include "components/metrics/single_sample_metrics.h"
 #include "components/viz/common/features.h"
@@ -634,6 +635,9 @@ void RenderThreadImpl::Init() {
   // been initialized by the Zygote before this instance became a Renderer.
   media::InitializeMediaLibrary();
 
+  memory_pressure_monitor_ = std::make_unique<memory_pressure::MultiSourceMemoryPressureMonitor>();
+  memory_pressure_monitor_->MaybeStartPlatformVoter();
+
   memory_pressure_listener_ = std::make_unique<base::MemoryPressureListener>(
       FROM_HERE,
       base::BindRepeating(&RenderThreadImpl::OnMemoryPressure,
diff --git a/content/renderer/render_thread_impl.h b/content/renderer/render_thread_impl.h
index 351210e73bde2da89645ad6c09e5c4820d2321b5..8d6f914ac2d499b78187b3c9216a50c02cf66034 100644
--- a/content/renderer/render_thread_impl.h
+++ b/content/renderer/render_thread_impl.h
@@ -86,6 +86,10 @@ namespace media {
 class GpuVideoAcceleratorFactories;
 }
 
+namespace memory_pressure {
+class MultiSourceMemoryPressureMonitor;
+}
+
 namespace mojo {
 class BinderMap;
 }
@@ -543,6 +547,8 @@ class CONTENT_EXPORT RenderThreadImpl
 
   HistogramCustomizer histogram_customizer_;
 
+  std::unique_ptr<memory_pressure::MultiSourceMemoryPressureMonitor> memory_pressure_monitor_;
+
   std::unique_ptr<base::MemoryPressureListener> memory_pressure_listener_;
 
   std::unique_ptr<viz::Gpu> gpu_;
